src/listwidget.cpp

```diff
@@ -2,233 +2,35 @@
 
 #include <algorithm>
 
-#include "utils.h"
-
 namespace newsboat {
 
 ListWidget::ListWidget(const std::string& list_name, Stfl::Form& form,
 	std::uint32_t scrolloff)
-	: list_name(list_name)
-	, form(form)
-	, num_lines(0)
-	, num_context_lines(scrolloff)
+	: backend(list_name, form)
+	, movement_control(backend, scrolloff)
 {
 }
 
-void ListWidget::stfl_replace_list(std::uint32_t number_of_lines,
-	std::string stfl)
+void ListWidget::stfl_replace_list(std::uint32_t number_of_lines, std::string stfl)
 {
-	num_lines = number_of_lines;
-	form.modify(list_name, "replace", stfl);
-
-	if (get_position() >= num_lines) {
-		if (num_lines > 0) {
-			set_position(num_lines - 1);
-		} else {
-			set_position(0);
-		}
-	}
+	backend.stfl_replace_list(number_of_lines, stfl);
+	movement_control.set_position(0);
```

Comment:
Shouldn't this call `movement_control.on_lines_changed()`?

Reply:
Oops, just missed this while applying+pushing the requested changes.

It is not really necessary, given that `movement_control.set_position(0)` will move both the position and scroll_offset to 0 (which is the old behavior).
However, I guess it makes sense to always call `on_lines_changed` to make sure this keeps working even if more internal state is added to the movement control